### PR TITLE
Feature/support list element validation for all constraints 

### PR DESCRIPTION
**Pull request by pelletier197**
Basically, every directive now has to specify if it supports list elements. If yes, instead of receiving the list of elements as an arguments, it receives each element individually. Which makes the validations easier to write, and gives better errors (specific index instead of global list error). 

The only thing with that solution is that two constraints didn't work well like that. `@NotEmpty` and `@Size` could work for both strings and lists. In order to avoid confusion and make it possible to support both constraints on an input, I added two new constraints. 
- `@ContainerSize` 
- `@ContainerNotEmpty`
which will only perform the validations on list inputs, while 
- `@Size`
- `@NotEmpty`
were modified to work only with strings, and IDs

I also simplified a little bit the directive code to reduce dupplication.

Resolves: #31

### CODE REVIEW

**Comment by pelletier197** [comment]
@bbakerman if you want to have a look at this. Sorry about it, it's quite large, but I think this is a good addition :wink:.

File: src/main/java/graphql/validation/constraints/standard/PatternConstraint.java [hunk 1]
```diff
@@ -51,40 +44,28 @@ public boolean appliesToType(GraphQLInputType inputType) {
     @Override
     protected List<GraphQLError> runConstraint(ValidationEnvironment validationEnvironment) {
         Object validatedValue = validationEnvironment.getValidatedValue();
-        GraphQLInputType argumentType = validationEnvironment.getValidatedType();
 
-        if (validatedValue == null) {
-            return emptyList();
-        }
```
Comment (bbakerman) [inline src/main/java/graphql/validation/constraints/standard/PatternConstraint.java:58]: Another another  of null being allowed

I dont recall exactly of hand out null handling policy but I am pretty sure we want graphql to catch null / non null lissues and these constraints o catch the rest

**Review by bbakerman — CHANGES_REQUESTED** [review]
Prove to me why the null value hanlding code (which returns true) is sensible to remove.

You have removed it all over the place.  Why is that ok?

**Comment by bbakerman** [comment]
I thank you for this PR - its extensive to say the least.  Smaller PRs usually have more success in open source terms.

I am musing on the breaking change here where @Size no longer does lists.  I still need to be convinced this is a good idea.

Also the null handling as outlined above

File: src/main/java/graphql/validation/constraints/standard/AbstractDecimalMinMaxConstraint.java [hunk 2]
```diff
@@ -1,60 +1,33 @@
...
+        return isOneOfTheseTypes(inputType, GRAPHQL_NUMBER_AND_STRING_TYPES);
     }
 
     @Override
     protected List<GraphQLError> runConstraint(ValidationEnvironment validationEnvironment) {
         Object validatedValue = validationEnvironment.getValidatedValue();
-        //null values are valid
-        if (validatedValue == null) {
-            return Collections.emptyList();
-        }
```
[the remark this answers is not in the document]
Comment (pelletier197) [reply src/main/java/graphql/validation/constraints/standard/AbstractDecimalMinMaxConstraint.java:57]: That's a good question. Here is why: 

https://github.com/graphql-java/graphql-java-extended-validation/blob/d779f85c728484887464393f45decf2423d3da13/src/main/java/graphql/validation/constraints/AbstractDirectiveConstraint.java#L120

The abstract parent class handles the null values and don't call the subclass methods in this case. This is not new with my changes actually, those if statements were never reached.

[the remark this answers is not in the document]
Comment (pelletier197) [reply src/main/java/graphql/validation/constraints/standard/AbstractDecimalMinMaxConstraint.java:57]: You can still see that null handling works in the tests.

**Comment by pelletier197** [comment]
I agree, I rarely do big PRs in open source. I tried to do too much at once.

Regarding the change of size, here's the issue I wanted to fix. When you have a schema like this. 

```graphql
field( arg : [ String ] @Size(min: 2)) : ID
```

It would have been possible to implement  `only the list should have a size of min 2` or `only the strings in the  list should have a size of min 2`,  but not both. And the user can't really choose which one he wants. 

Now, what I proposed is 
```graphql
field( arg : [ String ] @Size(min: 2) @ContainerSize(min : 5)) : ID
```

Which would say `only lists of at least 5 strings, and the strings must have at least a size of 2`.

Now maybe this change was not required. It may be an edge case, but it felt important to handle something like this. Or maybe there is another way I haven't seen?

**Comment by pelletier197** [comment]
@bbakerman to come back on this what to do you about my justification? If you don't like it, what do you propose?

And do you wish me to split this PR in smaller chunks?

**Comment by bbakerman** [comment]
Sorry for the delay on this PR - I have not been able to put the time towards this repo that it deserves.

I am happy to accept this PR and have the change in behavior

Thank you for such an extensive PR with great quality

**Comment by pelletier197** [comment]
Thanks a lot!

**Pull request merged**